lib/schema-compilers.js

```diff
@@ -0,0 +1,37 @@
+'use strict'
+
+const Ajv = require('ajv')
+const fastJsonStringify = require('fast-json-stringify')
+
+function ValidatorCompiler (externalSchemas, options, cache) {
+  // This instance of Ajv is private
+  // it should not be customized or used
+  const ajv = new Ajv(Object.assign({
+    coerceTypes: true,
+    useDefaults: true,
+    removeAdditional: true,
+    allErrors: true,
+    nullable: true
+  }, options.customOptions, { cache }))
+
+  if (options.plugins && options.plugins.length > 0) {
+    for (const plugin of options.plugins) {
+      plugin[0](ajv, plugin[1])
+    }
+  }
+
+  Object.values(externalSchemas).forEach(s => ajv.addSchema(s))
```

Comment:
While this is nice and will provide a really good encapsulation of the schemas, this will be horribly slow as it would add _all schemas_ to a new Ajv instance for every route. Note that addSchema performs a full validation of the added schema.

Reply:
This will be executed "only" for each fastify's instance/register and not for every route thanks to [this check](https://github.com/fastify/fastify/pull/2023/files#r371032778)

Do you agree if I try to arrange the code to create a new ajv instance only if the `externalSchemas` of the parent's instance isn't the same of the current that is being loaded?
It should works

Reply:
That'd be perfect, thanks.

Reply:
I should correct my sentence: the ajv instance is created only once per subtree:

- case 1: best, only one ajv instance is created
- case 2: worse, an ajv instance is created for every subtree (in yellow the ajv instances that I will try to optimize building only one)

![image](https://user-images.githubusercontent.com/11404065/74111698-ec4f0280-4b96-11ea-8ea4-d2cd4940cbe0.png)

Reply:
When is this function called?

Reply:
> That'd be perfect, thanks.

I added this feature. 
I will add more tests if you think my solution is sustainable (cause two `JSON.stringify`)

Reply:
Where are those JSON.stringify?

Reply:
[`lib/schema-compilers.js`](https://github.com/fastify/fastify/pull/2023/files#diff-3b3f7b51f94e63bb12088cdae2e0c9d3R9-R11)

I need a key to check if I should create a new AJV instance or not